### PR TITLE
[WIP] Spacemen are now Forklift Certified. Adds the Rapid Construction Forklifts to the game.

### PR DESCRIPTION
**Pull request by Iamgoofball**
## design doc

https://hackmd.io/@tgstation/rJjgKJEYc

## Early demo video:

https://user-images.githubusercontent.com/4081722/174664585-d4115a2a-9392-4ec7-9081-6a6fa360c767.mp4

## "How do I use this forklift?"

LMB to place. RMB to cancel/refund compatible item. Can't refund anything with contents or that isn't placable from that module.
Scroll wheel to select a new object from the current module.
Shift+Scrollwheel to change module.
Ctrl+Scrollwheel on most modules rotates (counter)clockwise.
Alt+Scrollwheel cycles through additional options, if available.
Middle Mouse also cycles through additional options, if available.

I'll finish the changelog and shit when it's done.

## TODO list:
- [x] Setup forklifts
- [x] Setup forklift modules
- [x] Setup construction holograms
- [x] Floors module
- [x] Walls module
- [x] Furniture module
- [x] Airlocks module
- [x] Departmental Machinery module
- [x] Shuttle module
- [x] Define departmental forklifts
- [x] Put keys in Head of Staff lockers
- [ ] Migrate to click catchers to avoid mouse entered, discuss w/ Potato
- [ ] Put keys in hands of highest hour count staffers in departments
- [ ] Find way to get more forklifts, cargo order maybe?
- [x] Put sprites in so it's not just the janicart

### CODE REVIEW

**Comment by Kitfox88** [comment]
Staplerfahrer Klaus time

File: code/modules/vehicles/forklift/modules/departmental_machinery.dm [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+			else
+				resource_price[initial(board_checked.build_path)] = machinery_price
+		continue
+
+/datum/forklift_module/department_machinery/on_scrollwheel(mob/source, atom/A, scrolled_up)
+	var/list/available_paths_real = list(
+		/obj/structure/frame/machine,
+		/obj/structure/frame/computer,
+	)
+	var/datum/techweb/station_techweb = SSresearch.techwebs_assoc_lookup[TECHWEB_STATION]
```
Comment (JohnFulpWillard) [inline code/modules/vehicles/forklift/modules/departmental_machinery.dm:48]: This is not good, the techweb should be stored locally and referred to there. This is hardcoding the machine to always use the station techweb, you never know if someone in the future would want to give one to Oldstation, Golems, Cybersun Ship, or whatever else exists in the future.

**Comment by SomeoneYouProbablyKnow** [comment]
AAAaaaa Scroll wheel use in my spessman game. 

Also that's really fast construction, what can i do if someone with a forklift tries to entomb me in walls and i wasn't paying attention for a second? With RCDs you gotta be close and  now you can break the construction thingy or move through it.

File: code/modules/vehicles/forklift/modules/departmental_machinery.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/vehicles/forklift/modules/departmental_machinery.dm:48]: oranges's design document does not specify support for oldstation/out of station locations, someone can patch that support in in the future

**Comment by Sakamoto4ka** [comment]
cant crush captain =(((

dreamseeker_x7kHLtzWH8

**Comment by deathrobotpunch** [comment]
since this got made into a new PR I’m going to ask again. are you going to be able to edit the hot keys the forklift uses?

**Comment by Iamgoofball** [comment]
the forklift is going to have a proper GUI coded by @ZephyrTFA once I'm done with the PR

File: code/modules/vehicles/forklift/modules/departmental_machinery.dm (hunk 1 above)
Comment (JohnFulpWillard) [reply code/modules/vehicles/forklift/modules/departmental_machinery.dm:48]: Oranges design document doesn't specify techwebs at all but it is bad to hardcode techwebs in general, I've been pushing to dehardcode them for a while and this is gonna regress it. Please change it.

**Comment by B4CKU** [comment]
it looks so cool!

**Comment by ZephyrTFA** [comment]
its not ready, gui isnt done

**Comment by Iamgoofball** [comment]
undrafting to see if it resolves an issue with TGS when trying to testmerge

**Comment by optimumtact** [comment]
you need to add a link to the design doc and a why it's good for the game .

**Comment by Admiral-Gigaton** [comment]
emag function: stabs people with prongs, causing wounds

**Comment by EEASAS** [comment]
your repo is a bit old dont you think

**Comment by optimumtact** [comment]
construction/deconstruction speeds need to be slower, people are saying it's too easy to grief.

Also it's too easy to steal the forklift right now, please make the key required for driving only, then the user takes the key out and has it in their pocket to use it's construction features.

**Comment by iprice** [comment]
Does this log anything anywhere?  Couldn't find it in PP or tools log, another obvious griefers tool, (hooray for RCD proliferation) and we only just getting to having working logging for RCDs.

Only 'log' I found in a quick search was from hologram, not actual logging.

If I just missed it please let me know where it gets logged.

**Comment by ilikesaying** [comment]
inb4 Forklift Safety Video recreated in TG

**Comment by Toastgoats** [comment]
If its too op, stick one that's locked to the shuttle in the BYOS. That would be fun

File: tgui/packages/tgui/interfaces/ConstructionForklift/constants.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const dir_num2str = (dir_num: 1 | 2 | 4 | 8) => {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/constants.tsx:1]: Just use an object. Switches aren't type safe (yes I realize how awful I sound)

File: tgui/packages/tgui/interfaces/ConstructionForklift/types.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
+type MaterialData = {
+  storage_max: number;
+  storage_now: number;
+  storage: Record<string, number>;
+};
+
+type AvailableModules = {
+  [ref: string]: string;
+};
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/types.tsx:9]: This is basically just record right? Just use Record<string, string>

**Comment by NotSomeoneyUKnow** [comment]
no fucking way

**Comment by Iamgoofball** [comment]
stale b gone

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        available_builds={active_module_data.available_builds}
+      />
+    );
+  }
+
+  const set_active_module = (module: string) => {
+    act('set-active', { new_module_ref: module });
+  };
+  let module_select = (
+    <ModuleSelect modules={modules} set_active_module={set_active_module} />
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:24]: just put these in the jsx please, no need to make them a var outside of it,

```tsx
<Window.Content>
{!!active_module_data && (<BuildTargetSelect
        selected={active_module_data.current_selected_typepath}
        available_builds={active_module_data.available_builds}
      />)}
<ModuleSelect modules={modules} set_active_module={() =>  act('set-active', { new_module_ref: module });} />
 </Window.Content>
 ```

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+};
+
+const ModuleAction = (action: string, params: object) => {
+  const { act } = useBackend();
+  const act_params = {
+    ...params,
+    action: action,
+  };
+  act('module-action', act_params);
+};
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:44]: not a fan of this, just call act directly in my opinion.

also this is not a component, no need to pascal case. components return JSX

`params: object` 
you're looking for `Record<string, any>`, objects, sadly, are not of type object

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+const BuildTargetSelect = (props: {
+  selected: string;
+  available_builds: BuildData;
+}) => {
+  const { selected, available_builds } = props;
+  const { act } = useBackend();
+  const select_item = (typepath: string) => {
+    ModuleAction('select-build-target', { typepath: typepath });
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:75]: just put act here if you can

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    display: 'inline-block',
+    cursor: 'pointer',
+    maxWidth: '32px',
+    maxHeight: '32px',
+  };
+
+  return (
+    <Image
+      src={props.display_src}
+      // @ts-ignore
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:142]: wut? prop doesnt exist, this error is intentional. do you want to make the tooltip here? 
`<Tooltip content={props.name}><Image ... /></Tooltip>`

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx (hunk 4 above)
Comment (ZephyrTFA) [reply tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:24]: I prefer it like this as it's cleaner and imo more legible

File: tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx (hunk 7 above)
Comment (ZephyrTFA) [reply tgui/packages/tgui/interfaces/ConstructionForklift/index.tsx:142]: this is using title, which is a builtin prop that exists on every html element.

**Comment by optimumtact** [comment]
you have failed to progress this yet again.

**Pull request closed without merging**